addons/binding/org.openhab.binding.smappee/src/main/java/org/openhab/binding/smappee/internal/SmappeeService.java

```diff
@@ -0,0 +1,473 @@
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link SamppeeService} handles the connection to the Smappee API
+ *
+ * @author Niko Tanghe - Initial contribution
+ */
+public class SmappeeService {
+
+    private final Logger logger = LoggerFactory.getLogger(SmappeeService.class);
+
+    ScheduledFuture<?> scheduledJob;
+
+    private int retry;
+
+    public SmappeeConfigurationParameters config;
+
+    private String serviceLocationId;
+
+    private boolean initialized;
+
+    private final Gson gson;
+
+    private String accessToken;
+    private DateTime accessTokenValidity;
+    private String refreshToken;
+
+    private static HttpClient httpClient = new HttpClient(new SslContextFactory());
+
+    public SmappeeService(SmappeeConfigurationParameters config) {
+        this.config = config;
+
+        this.retry = 1;
+
+        this.gson = new Gson();
+
+        this.initialized = false;
+
+        if (!httpClient.isStarted()) {
+            try {
+                httpClient.setFollowRedirects(false);
+                httpClient.start();
+            } catch (Exception e) {
+                logger.warn("Cannot start HttpClient!", e);
+            }
+        }
+
+    }
+
+    public void startAutomaticRefresh(ScheduledExecutorService scheduledExecutorService,
+            final ReadingsUpdate readingsUpdate) {
+        scheduledJob = scheduledExecutorService.scheduleWithFixedDelay(() -> {
+            SmappeeDeviceReading readings = getDeviceReadings();
+
+            // sometimes smappee returns reading without aggregated consumptions
+            if (readings != null && readings.consumptions.length > 0) {
+                readingsUpdate.newState(readings);
+            }
+        }, 0, config.poll_time, TimeUnit.MINUTES);
+    }
+
+    public void stopAutomaticRefresh() {
+        if (scheduledJob != null) {
+            scheduledJob.cancel(true);
+        }
+    }
+
+    /**
+     * Gets the smappee energy consumption readings.
+     *
+     * @return the device readings
+     */
+    public SmappeeDeviceReading getDeviceReadings() {
+        int currentTry = 0;
+        do {
+            try {
+                DateTime nowUtc = DateTime.now(DateTimeZone.UTC).minusMinutes(1);
+                DateTime nowUtcMinus20Min = DateTime.now(DateTimeZone.UTC).minusMinutes(20);
+
+                String nowUtcMillis = String.valueOf(nowUtc.getMillis());
+                String nowUtcMinus20MinMillis = String.valueOf(nowUtcMinus20Min.getMillis());
+
+                // sample API method to call :
+                // https://app1pub.smappee.net/dev/v1/servicelocation/123/consumption?aggregation=1&from=1388534400000&to=1391212800000
+
+                String responseReadings = getData("/dev/v1/servicelocation/" + this.serviceLocationId
```

Comment:
"/dev/v1/servicelocation/" would be an excellent candidate for a constant, wouldn't it?

Reply:
Yes, it's a part of an url that happens to be same for several REST endpoints. putting it in a constant will not make the code more readable or more performant.

Reply:
No, but it will make refactorings less error-prone, e.g. if this url segment cahnged to a `v2` in future...